NUTCH-2375 Upgrade the code base from org.apache.hadoop.mapred to org.apache.hadoop.mapreduce

Omkar20895 opened the pull request:
- This PR is a part of the upgrade and will be updated continuously by me.
- Please feel free to review the PR.

lewismc commented:
Excellent @Omkar20895 thank you for starting this pull request. Now you will see that EVERY class is broken e.g. does not compile... lets gradually fix those classes. Please update this PR as you progress. Thank you for submitting PR early it makes a huge difference for review.

lewismc commented:
@Omkar20895 please keep the updates coming as we can review the code incrementally. Thanks.

Omkar20895 commented:
@lewismc I am working on upgrading crawldb, I will send the update as soon as I am done with it. Thanks.

Omkar20895 commented:
CrawlDb needs to be updated more, this will be done in the coming weeks. Major changes still need to be done in DeDuplicationJob.java, Thanks. 
@lewismc

Omkar20895 commented:
@lewismc please review the latest commit so that I can incorporate the changes suggested by you.

lewismc commented on src/java/org/apache/nutch/crawl/CrawlDb.java:
```diff
@@ -28,8 +28,9 @@
 import org.apache.hadoop.conf.*;
-import org.apache.hadoop.mapred.*;
 import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.lib.input.*;
```
Please never use the wildcard imports e.g. ```import org.apache.hadoop.mapreduce.lib.input.*;```, always make individual imports.

Omkar20895 replied:
I am changing it in the local copy and I will update the PR today.

lewismc replied:
Can you please update this as well?

lewismc commented on src/java/org/apache/nutch/crawl/CrawlDbMerger.java:
```diff
@@ -32,7 +32,9 @@
 import org.apache.hadoop.io.Text;
 import org.apache.hadoop.io.Writable;
-import org.apache.hadoop.mapred.*;
+import org.apache.hadoop.mapreduce.*;
```
Same here.

lewismc replied:
@Omkar20895 please address the above issue

lewismc commented on src/java/org/apache/nutch/crawl/CrawlDbReader.java:
```diff
@@ -47,22 +47,10 @@
-import org.apache.hadoop.mapred.lib.HashPartitioner;
-import org.apache.hadoop.mapred.lib.IdentityMapper;
-import org.apache.hadoop.mapred.lib.IdentityReducer;
+import org.apache.hadoop.mapreduce.*;
```
Same here.

lewismc commented on src/java/org/apache/nutch/crawl/CrawlDbReader.java:
```diff
@@ -371,26 +361,27 @@ public void close() {
 
-	  JobConf job = new NutchJob(config);
+	  Job job = new NutchJob(config);
+          config = job.getConfiguration();
```
Formatting.

lewismc commented on src/java/org/apache/nutch/crawl/DeduplicationJob.java:
```diff
@@ -31,7 +31,7 @@
 import org.apache.hadoop.io.BytesWritable;
 import org.apache.hadoop.io.Text;
-import org.apache.hadoop.mapred.Counters.Group;
+/*import org.apache.hadoop.mapred.Counters.Group;
```
Never leave code commented out like this please. Also make explicit individual imports please.

lewismc commented on src/java/org/apache/nutch/crawl/Generator.java:
```diff
@@ -29,8 +29,10 @@
 import org.apache.hadoop.conf.*;
-import org.apache.hadoop.mapred.*;
-import org.apache.hadoop.mapred.lib.MultipleSequenceFileOutputFormat;
+import org.apache.hadoop.mapreduce.*;
```
Same here.

lewismc commented on src/java/org/apache/nutch/util/NutchJob.java:
```diff
@@ -17,14 +17,16 @@
-  public NutchJob(Configuration conf) {
-    super(conf, NutchJob.class);
+  public NutchJob(Configuration conf) throws IOException {
+    super(conf, "NutchJob");
```
Does this mean that every Job will be named ```"NutchJob"```?

Omkar20895 replied:
yes, @lewismc if we create a job as NutchJob job = new NutchJob(); then it will be named by default NutchJob and it can be overriden as job.setJobName("<name>"); in the code where the job is being created. Should it be in any other way?

Omkar20895 commented:
Please ignore the number of commits, I will squash them while this PR is being merged. Thanks.

lewismc commented:
np @Omkar20895 keep them coming :)

lewismc commented:
Also @Omkar20895 please ensure that this PR is kept up-to-date with master branch or else we will end up in mess. Thank you

Omkar20895 commented:
yes I will @lewismc

Omkar20895 commented:
@lewismc please review the latest commit. The errors I was mentioning will be found in the files crawl/Generator.java, crawl/CrawlDbReader.java, and crawl/DeduplicationJob.java

lewismc commented:
@Omkar20895 I have no idea what the errors are as I am getting a whole host of compile issues due to the changing nature of NutchJob with the upgrade. 
Please go ahead and upgrade all instances of 
```
JobConf job = new NutchJob(getConf());
```
We can then take it job-by-job, thanks. Also please make the updates as per my previous review and comments. Let's keep this branch up-to-date with master and also up-to-date with my comments. Thank you very much @Omkar20895

lewismc commented:
@Omkar20895 you will also see the following
```
/usr/local/nutch/src/java/org/apache/nutch/util/NutchJob.java:29: warning: [deprecation] Job(Configuration,String) in Job has been deprecated
    [javac]     super(conf, "NutchJob");
```
This also needs to be fixed before you go on and do anything else.

Omkar20895 commented:
Hi, @lewismc this concludes updating of all files in nutch codebase. There are a couple of repetitive errors that I could not solve. Please feel free to review the PR and also the errors. Thanks.

lewismc commented on src/java/org/apache/nutch/crawl/CrawlDb.java:
```diff
@@ -28,8 +28,12 @@
 import org.apache.hadoop.io.*;
```
@Omkar20895 can you please sort out all instances of wildcard imports for Hadoop?

Omkar20895 commented:
Hi @lewismc , the major is error is produced by SequenceFileOutputFormat.getReaders() which has been removed from the latest mapreduce API, I cannot find the replacement for it. Thanks.

Omkar20895 commented:
Hi @lewismc, did you get a chance to look at this issue? Thanks.

lewismc commented:
Yes I did @Omkar20895

lewismc commented:
I'll comment on thr PR

lewismc commented:
Hi @Omkar20895 I am getting the following build errors
```
compile-core:
    [javac] Compiling 284 source files to /usr/local/nutch/build/classes
    [javac] /usr/local/nutch/src/java/org/apache/nutch/crawl/CrawlDb.java:119: error: unreported exception InterruptedException; must be caught or declared to be thrown
    [javac]       int complete = job.waitForCompletion(true)?0:1;
    [javac]                                           ^
    [javac] /usr/local/nutch/src/java/org/apache/nutch/crawl/CrawlDbReader.java:399: error: cannot find symbol
    [javac] 	  SequenceFile.Reader[] readers = SequenceFileOutputFormat.getReaders(config,
    [javac] 	                                                          ^
    [javac]   symbol:   method getReaders(Configuration,Path)
    [javac]   location: class SequenceFileOutputFormat
    [javac] /usr/local/nutch/src/java/org/apache/nutch/hostdb/ReadHostDb.java:208: error: cannot find symbol
    [javac]     SequenceFile.Reader[] readers = SequenceFileOutputFormat.getReaders(conf, hostDb);
    [javac]                                                             ^
    [javac]   symbol:   method getReaders(Configuration,Path)
    [javac]   location: class SequenceFileOutputFormat
    [javac] /usr/local/nutch/src/java/org/apache/nutch/segment/SegmentReader.java:434: error: cannot find symbol
    [javac]     SequenceFile.Reader[] readers = SequenceFileOutputFormat.getReaders(
    [javac]                                                             ^
    [javac]   symbol:   method getReaders(Configuration,Path)
    [javac]   location: class SequenceFileOutputFormat
    [javac] /usr/local/nutch/src/java/org/apache/nutch/segment/SegmentReader.java:511: error: cannot find symbol
    [javac]       SequenceFile.Reader[] readers = SequenceFileOutputFormat.getReaders(
    [javac]                                                               ^
    [javac]   symbol:   method getReaders(Configuration,Path)
    [javac]   location: class SequenceFileOutputFormat
    [javac] /usr/local/nutch/src/java/org/apache/nutch/tools/FreeGenerator.java:197: error: cannot find symbol
    [javac]     job.setOutputKeyComparatorClass(Generator.HashComparator.class);
    [javac]        ^
    [javac]   symbol:   method setOutputKeyComparatorClass(Class<HashComparator>)
    [javac]   location: variable job of type Job
    [javac] /usr/local/nutch/src/java/org/apache/nutch/util/NutchJob.java:29: warning: [deprecation] Job(Configuration,String) in Job has been deprecated
    [javac]     super(conf, "NutchJob");
    [javac]     ^
    [javac] Note: Some input files use unchecked or unsafe operations.
    [javac] Note: Recompile with -Xlint:unchecked for details.
    [javac] 6 errors
    [javac] 1 warning
```

lewismc commented:
Please start with the following issue
```
    [javac]   symbol:   method setOutputKeyComparatorClass(Class<HashComparator>)
    [javac]   location: variable job of type Job
    [javac] /usr/local/nutch/src/java/org/apache/nutch/util/NutchJob.java:29: warning: [deprecation] Job(Configuration,String) in Job has been deprecated
    [javac]     super(conf, "NutchJob");
    [javac]     ^
```

lewismc commented:
@Omkar20895 please refer to following class for guidance
https://github.com/apache/nutch/blob/2.x/src/java/org/apache/nutch/util/NutchJob.java
Once you push a commit please ping me. Thank you.

Omkar20895 commented:
@lewismc please take a look at this commit.

Omkar20895 commented on src/java/org/apache/nutch/util/NutchJob.java:
```diff
@@ -25,8 +25,9 @@
-  public NutchJob(Configuration conf) throws IOException {
-    super(conf, "NutchJob");
-  }
+  public static NutchJob getJobInstance(Configuration conf){
```
This latest addition will still give error because we do not have an constructor in this class and JVM will add a default constructor which will have a default super call to the parent class. Something of this form : 
***default constructor***
public NutchJob(){
  super()
}

But this won't create any discrepancy because we are not utilizing that default constructor in the whole code base.

Omkar20895 replied:
This doesn't give an error but gives a warning.

Omkar20895 commented:
Hi, @lewismc any update on this?

lewismc commented:
Hi @Omkar20895 please update all instance of ```SequenceFile.Reader[] readers = SequenceFileOutputFormat.getReaders(getConf(), dir);``` to ```MapFile.Reader[] readers = MapFileOutputFormat.getReaders(dir, getConf());```

Omkar20895 commented:
But mapfileoutputformat and sequencefileoutputformat are different aren't they? Won't it cause some discrepancy? 
@lewismc

lewismc commented:
> But mapfileoutputformat and sequencefileoutputformat are different aren't they?

Yes they have different semantics. You can see difference between [MapFile](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/io/MapFile.html) and [SequenceFile](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/io/SequenceFile.html).

> Won't it cause some discrepancy?

It may well do. Right now though you can implement them (there are only 4 or 5 instances) we have this thread for reference. If there are discrepancies then we can come back and fix them.
\Most import thing is that you continue to move onwards.

Omkar20895 commented:
Done! what about the other two errors.
One is I cannot find the replacement for
/home/omkar/Documents/GIT/nutch1/src/java/org/apache/nutch/tools/FreeGenerator.java:197: error: cannot find symbol
    [javac]     job.setOutputKeyComparatorClass(Generator.HashComparator.class);

@lewismc

lewismc commented:
@Omkar20895 before I look at this, which Class or Method is no longer found? ```setOutputKeyComparatorClass```, ```Generator```, or ```HashComparator```?

Omkar20895 commented:
@lewismc setOutputKeyComparatorClass.

lewismc commented:
@Omkar20895 did you push your new commits? I don't see them yet and I just tried pulling.

lewismc commented:
Please replace ```job.setOutputKeyComparatorClass``` with ```job.setSortComparatorClass```

Omkar20895 commented:
`public class NutchJob extends JobConf `
Okay let me tell you my why I changed it, you see JobConf is no longer there in new Mapreduce API and hence I replaced it with Job class and now NutchJob extends Job class. 
```
public NutchJob(Configuration conf) {
    super(conf, NutchJob.class);
}
```
If you look at the constructor it has a super call that will call Job(conf, class) which is deprecated and the documentation suggests to use getInstance instead. 

Please refer to constructor summary section [here](https://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/mapreduce/Job.html).  

Please feel free to correct me if I am wrong. Thanks.

lewismc commented:
OK, just as long as the deprecation WARN is removed at some stage I am happy with that. Looking forward to this upgrade finishing and we can test it out. Thank you @Omkar20895

Omkar20895 commented:
Hi there is this one last error that I am concerned about. 

```
[javac] /home/omkar/Documents/GIT/nutch1/src/java/org/apache/nutch/crawl/CrawlDb.java:119: error: unreported exception InterruptedException; must be caught or declared to be thrown
    [javac]       int complete = job.waitForCompletion(true)?0:1;
```
I have used `int complete = job.waitForCompletion(true)?0:1;` in multiple files and I did not get any error in those files and I am unable to understand why I am getting this here. Thanks.

lewismc commented:
@Omkar20895 ... again, you've not pushed anything to the remote branch. Please push so I can pull locally. Thanks.

Omkar20895 commented:
@lewismc pardon me for the delay. I thought I pushed a commit. 😅

lewismc commented:
Check out [Job.waitForCompletion](https://hadoop.apache.org/docs/current/api/index.html?org/apache/hadoop/mapreduce/Job.html) it
```
Throws:
    IOException - thrown if the communication with the JobTracker is lost
    InterruptedException
    ClassNotFoundException
```
These need to be handled appropriately, please go and have a look at Exception handling within method calls e.g. try... catch

Omkar20895 commented:
@lewismc 

>>>`I have used int complete = job.waitForCompletion(true)?0:1; in multiple files and I did not get any error in those files and I am unable to understand why I am getting this here`

This exception handling needs to be done in every file that used the above line, I am working on it and will push a commit as soon as I am done. I am making following changes.

Converting `int complete = job.waitForCompletion(true)?0:1;` to 

```
try {
      int complete = job.waitForCompletion(true)?0:1;
    } catch (InterruptedException e) {
      LOG.info("Exception: "+e);
      throw e;
    } catch (ClassNotFoundException e) {
      LOG.info("Exception: "+e);
      throw e;
    }
```
Thanks.

sebastian-nagel commented:
Hi @Omkar20895, blindly catching exceptions (logging and throwing them again) does not make the code better. It's important to think whether the exception needs to be handled or not If the job job.waitForCompletion(true) throws an exception the job has failed
- sometimes you need to clean up to avoid that long-living data structures (CrawlDb, LinkDb) are broken, e.g. in [Injector](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/crawl/Injector.java#L402)
- for write-once structures (segments) this is usually not a requirement, as broken segments are just ignored by other tools. In this case it's enough to throw the exception (it needs to be declared to be thrown)
Afaics, you're on the right way. Just make sure before a push that everything compiles (`ant clean runtime javadoc test`).

sebastian-nagel commented on src/java/org/apache/nutch/crawl/CrawlDb.java:
```diff
@@ -111,7 +116,7 @@ public void update(Path crawlDb, Path[] segments, boolean normalize,
     try {
-      JobClient.runJob(job);
+      int complete = job.waitForCompletion(true)?0:1;
     } catch (IOException e) {
```
Need to catch all thrown exceptions, incl. ClassNotFoundException and InterruptedException. Could also just catch `Exception`.

Omkar20895 replied:
@sebastian-nagel in the above job when there is an IOException the output path is being removed (cleaning up) before throwing up the error and the same needs to be done when we catch the other two exceptions i.e ClassNotFoundException and InterruptedException also, isn't it? Please feel free to correct me if I am wrong.

sebastian-nagel replied:
Yes, exactly, e.g.
```
public void update(...) throws Exception {
  ...
    try {
        job.waitForCompletion(true);
    } catch (IOException | ClassNotFoundException | InterruptedException e) {
       // do cleanup
      throw e
```

sebastian-nagel replied:
I left `complete` away because it is not used.

sebastian-nagel replied:
Of course, you could also (maybe a matter of taste):
```
public void update(...) throws IOException, ClassNotFoundException, InterruptedException {
```

Omkar20895 commented:
The next commit will be updating the plugins and tests which I am working on. I will update the PR as soon as I am done. Thanks.

lewismc commented:
Thank you @sebastian-nagel excellent advice

lewismc commented:
How are things coming along @Omkar20895 ?

Omkar20895 commented:
Hi,
These are the warnings and errors that I am getting right now and unable to solve. 

```
compile-core-test:
    [javac] Compiling 54 source files to /Users/omkar/Documents/Git/nutch/build/test/classes
    [javac] /Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/crawl/CrawlDbUpdateTestDriver.java:95: warning: [deprecation] setConfiguration(Configuration) in TestDriver has been deprecated
    [javac]     reduceDriver.setConfiguration(configuration);
    [javac]                 ^
    [javac] /Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/crawl/CrawlDbUpdateUtil.java:58: error: cannot find symbol
    [javac]     reducer.configure(Job.getInstance(conf));
    [javac]            ^
    [javac]   symbol:   method configure(Job)
    [javac]   location: variable reducer of type T
    [javac]   where T is a type-variable:
    [javac]     T extends Reducer<Text,CrawlDatum,Text,CrawlDatum> declared in class CrawlDbUpdateUtil
```
The reducer we are referring to here is [CrawlDbReducer](https://github.com/apache/nutch/blob/3e2d3d456489bf52bc586dae0e2e71fb7aad8fe7/src/java/org/apache/nutch/crawl/CrawlDbReducer.java) and even though it has a configure method in it I am getting the error.

```
    [javac] /Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/crawl/CrawlDbUpdateUtil.java:62: error: CrawlDbUpdateUtil.DummyContext is not abstract and does not override abstract method write(Object,Object) in TaskInputOutputContext
    [javac]   private class DummyContext extends Context {
    [javac]           ^
    [javac] 
```
Here I have already overrided the write method but still I am getting this error.

```
/Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/crawl/CrawlDbUpdateUtil.java:124: error: reduce(KEYIN,Iterable<VALUEIN>,Reducer<KEYIN,VALUEIN,KEYOUT,VALUEOUT>.Context) has protected access in Reducer
    [javac]       reducer.reduce(dummyURL, (Iterable)values, context);
    [javac]              ^
    [javac]   where KEYIN,VALUEIN,KEYOUT,VALUEOUT are type-variables:
    [javac]     KEYIN extends Object declared in class Reducer
    [javac]     VALUEIN extends Object declared in class Reducer
    [javac]     KEYOUT extends Object declared in class Reducer
    [javac]     VALUEOUT extends Object declared in class Reducer
    [javac] /Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/indexer/TestIndexerMapReduce.java:172: warning: [deprecation] setConfiguration(Configuration) in TestDriver has been deprecated
    [javac]     reduceDriver.setConfiguration(configuration);
    [javac]                 ^
    [javac] Note: /Users/omkar/Documents/Git/nutch/src/test/org/apache/nutch/crawl/CrawlDbUpdateUtil.java uses unchecked or unsafe operations.
    [javac] Note: Recompile with -Xlint:unchecked for details.
    [javac] 3 errors
    [javac] 2 warnings

```
Please pull and try out the latest commit to reproduce the errors and use the command "ant clean runtime test". The runtime build is successful but the test will give errors. Thanks.

lewismc commented:
@Omkar20895 

regarding ```CrawlDbUpdateTestDriver.java:95: warning: [deprecation] setConfiguration(Configuration) in TestDriver has been deprecated```
please see [javadoc](https://mrunit.apache.org/documentation/javadocs/1.1.0/index.html?org/apache/hadoop/mrunit/TestDriver.html): Deprecated. Use getConfiguration() to set configuration items as opposed to overriding the entire configuration object as it's used internally. Also make sure to update the Class and Method-level API documentation to ```{@link CrawlDbReducer#reduce(Text, Iterator, Context)}``` this is present in several Classes and you have not updated it yet.

regarding ```CrawlDbUpdateUtil.java:58: error: cannot find symbol``` this is due to removal of ```configure``` method for Reducer.java. Please see [Javadoc](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/mapreduce/Reducer.html). You may have to work with ```setup``` instead.

Fix both of the above, which once addressed will bring the compilation back to stable I think. Also check out the Javadoc for [TaskInputOutputContext](https://hadoop.apache.org/docs/current/api/index.html?org/apache/hadoop/mapreduce/TaskInputOutputContext.html) and ensure that CrawlDbUpdateUtil.DummyContext overrides the appropriate methods with the appropriate peramaters.

lewismc commented:
Thank you for updating PR @Omkar20895 , please begin making your way through the tests and fixing the errors. AFAICT, the following tests currently fail. You can see all of the reports in ```build/tests```

```
    [junit] Running org.apache.nutch.crawl.TestCrawlDbFilter
    [junit] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.099 sec
    [junit] Test org.apache.nutch.crawl.TestCrawlDbFilter FAILED
    [junit] Running org.apache.nutch.crawl.TestCrawlDbMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.986 sec
    [junit] Test org.apache.nutch.crawl.TestCrawlDbMerger FAILED
    [junit] Running org.apache.nutch.crawl.TestGenerator
    [junit] Tests run: 4, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 8.483 sec
    [junit] Test org.apache.nutch.crawl.TestGenerator FAILED
    [junit] Running org.apache.nutch.crawl.TestLinkDbMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.448 sec
    [junit] Test org.apache.nutch.crawl.TestLinkDbMerger FAILED
    [junit] Running org.apache.nutch.fetcher.TestFetcher
    [junit] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.323 sec
    [junit] Test org.apache.nutch.fetcher.TestFetcher FAILED
    [junit] Running org.apache.nutch.indexer.TestIndexerMapReduce
    [junit] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.766 sec
    [junit] Test org.apache.nutch.indexer.TestIndexerMapReduce FAILED
    [junit] Running org.apache.nutch.plugin.TestPluginSystem
    [junit] Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.073 sec
    [junit] Test org.apache.nutch.plugin.TestPluginSystem FAILED
    [junit] Running org.apache.nutch.segment.TestSegmentMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 23.399 sec
    [junit] Test org.apache.nutch.segment.TestSegmentMerger FAILED
    [junit] Running org.apache.nutch.segment.TestSegmentMergerCrawlDatums
    [junit] Tests run: 7, Failures: 7, Errors: 0, Skipped: 0, Time elapsed: 32.267 sec
    [junit] Test org.apache.nutch.segment.TestSegmentMergerCrawlDatums FAILED
```

@Omkar20895 have you tried to execute an end-to-end test crawl and validate the results? If not, then start there and see where it is all going wrong.
Thanks

lewismc commented on src/java/org/apache/nutch/segment/SegmentMerger.java:
```diff
@@ -384,40 +385,43 @@ public void setConf(Configuration conf) {
     setConf(conf);
     if (sliceSize > 0) {
-      sliceSize = sliceSize / conf.getNumReduceTasks();
+      sliceSize = sliceSize / Integer.parseInt(conf.get("mapreduce.map.tasks"));
```
This does not look correct at all... we should be attempting to obtain a slice size value by diving an integer whose value is greater than zero by the configured number of Reduce tasks... not by the configured number of Map tasks.

Omkar20895 replied:
@lewismc pardon me, my bad! I will make this change and commit it in the next update.

lewismc commented on src/java/org/apache/nutch/segment/SegmentMerger.java:
```diff
@@ -384,40 +385,43 @@ public void setConf(Configuration conf) {
+      String url = key.toString();
+      if (normalizers != null) {
+        try {
+          url = normalizers.normalize(url, URLNormalizers.SCOPE_DEFAULT); // normalize
```
Sort the comment out, make sure everything is on the same line.

Omkar20895 replied:
Done!

lewismc commented:
@jnioche @sebastian-nagel @chrismattmann @jorgelbg and rest of team, would it be preferred for all Mapper and Reducer functions to be extracted from individual Job's which ```extends Configured implements Tool``` into a ```mapper``` and ```reducer``` package respectively or is it fine to retain them as nested classes within the Job they belong to? Thanks for any input.

lewismc commented:
@Omkar20895 start with ```TestSegmentMergerCrawlDatums.java```, the primary issue is as follows
```
2017-07-08 17:27:43,515 WARN  mapred.LocalJobRunner (LocalJobRunner.java:run(560)) - job_local1442546213_0007
java.lang.Exception: java.lang.ClassCastException: org.apache.nutch.crawl.CrawlDatum cannot be cast to org.apache.nutch.metadata.MetaWrapper
	at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:462)
	at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:522)
Caused by: java.lang.ClassCastException: org.apache.nutch.crawl.CrawlDatum cannot be cast to org.apache.nutch.metadata.MetaWrapper
	at org.apache.nutch.segment.SegmentMerger$SegmentMergerMapper.map(SegmentMerger.java:397)
	at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:146)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:787)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341)
	at org.apache.hadoop.mapred.LocalJobRunner$Job$MapTaskRunnable.run(LocalJobRunner.java:243)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Omkar20895 commented:
@lewismc working on it! Thanks.

Omkar20895 commented:
@lewismc I think we need to replace the configure() method with setup(). What would you suggest? Thanks for any suggestion.

lewismc commented:
@Omkar20895 yes

Omkar20895 commented:
@lewismc Thanks, I will start working on it.

Omkar20895 commented:
I am working on stabilizing the tests for CrawlDbMerger and will commit it once I am done. Thanks.
@lewismc

lewismc commented:
OK good @Omkar20895 let me know by usual communication channels if you are struggling at all. Thanks.

lewismc commented:
I'm still seeing the following failures in test
```
    [junit] Running org.apache.nutch.crawl.TestCrawlDbMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.095 sec
    [junit] Test org.apache.nutch.crawl.TestCrawlDbMerger FAILED
    [junit] Running org.apache.nutch.crawl.TestGenerator
    [junit] Tests run: 4, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 9.021 sec
    [junit] Test org.apache.nutch.crawl.TestGenerator FAILED
    [junit] Running org.apache.nutch.crawl.TestLinkDbMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.524 sec
    [junit] Test org.apache.nutch.crawl.TestLinkDbMerger FAILED
    [junit] Running org.apache.nutch.fetcher.TestFetcher
    [junit] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.788 sec
    [junit] Test org.apache.nutch.fetcher.TestFetcher FAILED
    [junit] Running org.apache.nutch.indexer.TestIndexerMapReduce
    [junit] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.002 sec
    [junit] Test org.apache.nutch.indexer.TestIndexerMapReduce FAILED
    [junit] Running org.apache.nutch.plugin.TestPluginSystem
    [junit] Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.016 sec
    [junit] Test org.apache.nutch.plugin.TestPluginSystem FAILED
    [junit] Running org.apache.nutch.segment.TestSegmentMerger
    [junit] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 24.588 sec
    [junit] Test org.apache.nutch.segment.TestSegmentMerger FAILED
    [junit] Running org.apache.nutch.segment.TestSegmentMergerCrawlDatums
    [junit] Tests run: 7, Failures: 7, Errors: 0, Skipped: 0, Time elapsed: 35.151 sec
    [junit] Test org.apache.nutch.segment.TestSegmentMergerCrawlDatums FAILED
```
I'm sure you are aware of these. Let's hash them out offline.

lewismc commented:
hi @Omkar20895 please look at the following document for ideas on how comparisons in Java + JUnit should be done and what the impacts of different methods is.
https://stackoverflow.com/questions/27605714/test-two-instances-of-object-are-equal-junit
Please write back here or else on GChat if you re still struggling. We ned to move on and begin coding the new GraphGenerator tool.

Omkar20895 commented:
Fix for the rest of the tests on the way. Thanks.

Omkar20895 commented on src/java/org/apache/nutch/segment/SegmentMerger.java:
```diff
@@ -174,47 +174,51 @@
+            splitReader.initialize(split, context);
+            this.initialize(split,context);
+            boolean res = splitReader.nextKeyValue();
+            wrapper = this.getCurrentValue();
```
Hi @lewismc I think the TestSegmentMerger is failing here with a NullPointerException and I think that it is because the sequenceFileRecordReader here is not being intialized as expected. Please correct me if I am wrong. Thanks.

Omkar20895 commented:
Hi, I believe that all the changes in this PR are stable now. Please feel free to review it and I would be glad if I can receive any suggestions. Thanks.

lewismc commented:
@Omkar20895 please squash your commits and we will undertake community peer review thank you.

lewismc commented:
@Omkar20895 ping

Omkar20895 commented:
@lewismc Please find above the squashed commit. Thanks.

lewismc commented:
Hi @Omkar20895 you've not squashed the commits... if you had squashed them it would say 

> Omkar20895  wants to merge 1 commits into apache:master from Omkar20895:NUTCH-2375

not

> Omkar20895  wants to merge 46 commits into apache:master from Omkar20895:NUTCH-2375

Can you please try squashing again? Thanks
